File: solutions/0791. Custom Sort String/0791.cpp
// Time complexity: O(∣order∣+∣s∣)
// Space complexity: O(26)
class Solution {
public:
  string customSortString(string_view order, string_view s) {
    string ans;
    vector<int> count(128);

    for (const char c : s)
      ++count[c];

    for (const char c : order)
      while (count[c]-- > 0)
        ans += c;

    for (char c = 'a'; c <= 'z'; ++c)
      while (count[c]-- > 0)
        ans += c;

    return ans;
  }
};

File: solutions/0791. Custom Sort String/0791.js
// Time complexity: O(∣order∣+∣s∣)
// Space complexity: O(26)
var customSortString = function(order, s) {
  let ans = '';
  const count = Array(26).fill(0);

  for (const c of s)
    ++count[c.charCodeAt(0) - 'a'.charCodeAt(0)];

  for (const c of order) {
    ans += c.repeat(count[c.charCodeAt(0) - 'a'.charCodeAt(0)]);
    count[c.charCodeAt(0) - 'a'.charCodeAt(0)] = 0;
  }
      
  for (let i = 0; i < 26; i++)
    ans += String.fromCharCode('a'.charCodeAt(0) + i).repeat(count[i]);

  return ans;
};